semantic/solver.go

```diff
@@ -0,0 +1,61 @@
+package semantic
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+)
+
+type SolutionMap map[Node]Type
+
+func SolveTypes(program *Program, tenv TypeEnvironment, constraints ConstraintSet) (SolutionMap, error) {
+	solution := make(SolutionMap)
+	log.Println("tenv", tenv)
+	log.Println("constraints", constraints)
+	hasFreeVars := true
+	for hasFreeVars {
+		hasFreeVars = false
+		for i, a := range constraints {
+			for j, b := range constraints {
+				if i == j {
+					continue
+				}
+				constraints[j] = b.Substitute(a).(Constraint)
```

Comment:
What if you have two constraints with the same type variable on the left hand side:

* `tx` = substitutable type expression
* `tx` = a different substitutable type expression

And `tx` does not appear on the right hand side of any constraint? I think there needs to be substitutions on the left hand side of constraints as well.